### ISSUE TITLE
Condition#should_retry invoked more than expected

### ISSUE THREAD

**Issue by softprops** [srijs/rust-futures-backoff#1]
Thanks for this library. I've taken a look at a number of retry libraries for rust futures and this by far has the most straightforward and well thought out interface.

I'm working this into a client library and am trying to implement a `Condition` type that debug logs how many time an operation was retried. Here's a sketch

```rust
/// Predicate trait that determines if an impl
/// type is retryable
trait Retry {
    /// Return true if type is retryable
    fn retryable(&self) -> bool;
}

struct Counter(u16);

impl<R> Condition<R> for Counter
where
    R: Retry,
{
    fn should_retry(
        &mut self,
        error: &R,
    ) -> bool {
        log::debug!("retrying operation {}", self.0);
        self.0 += 1;
        error.retryable()
    }
}
```

```rust
Strategy::exponential(duration)
  .with_max_retries(5)
  .with_jitter(true)
  .retry_if(
    operation,
    Counter(0)
)
```

when I try a forced failing operation, I see some suprising output.

```
retrying operation 0
retrying operation 0
retrying operation 1
retrying operation 2
retrying operation 3
retrying operation 4
retrying operation 5
```

I'm wondering if this is expected.

What I expected was that the condition would be invoked once per future failure.

**Comment by srijs** [comment]
Generally yes, the condition should be invoked once per failure. I'm a bit puzzled about the output you posted, does it really print `0` twice? Intuitively I'd say there is no way for `futures-backoff` to invoke your condition in a way that it would print `0` two times in a row.

**Comment by softprops** [comment]
It does :/ here's a more complete example https://github.com/softprops/dynomite/blob/master/dynomite/src/retry.rs#L137 

I tried switching to an atomicusize briefly thinking it might be a thread visibly thing but observed the same behavior.

Ill try to put together a more focused example. Some addition thoughts are the additional wrapper to conform to a risotto type may somehow have an unexpected effect. The double 0 thing is very odd.

I opened an issue mainly to throw a line out incase I was misunderstanding. I'll close for now until I can make a smaller more focused example
